### ISSUE TITLE
Bulk import of Mongoid documents incorrect on Tire 0.5.1

### ISSUE THREAD

**Issue by digitalplaywright** [karmi/retire#529]
Bulk import of Mongoid documents in Tire 0.5.1 creates documents in the index with an erroneous id. The document id in the ES index is assigned a random string id instead of the bson object id. The same error may also be present on new document creation, but it seems to happen quite seldomly and  we do not have a testcase that demonstrates this. 

This example demonstrates the problem:

```
class TestCase
  include Mongoid::Document
  include Tire::Model::Search
  include Tire::Model::Callbacks

  cattr_reader :per_page
  @@per_page = 10

  def self.paginate(options = {})
    page(options[:page]).per(options[:per_page])
  end

  def self.build_new
    TestCase.destroy_all
    TestCase.tire.index.delete

    for i in 1..100
      tk = TestCase.new(:name => "foobar")
      tk.save!
    end

    TestCase.tire.index.delete
    TestCase.create_elasticsearch_index
    TestCase.tire.import
  end

  field :name, :type => String

  mapping do
    indexes :name,  :type => 'string'
  end

end
```

We use Mongoid 3.0.1, Moped 1.3.0 and Origin 1.0.11. We use Elastic Seach 0.19.11. Please let us know if you are able to reproduce the problem.

**Comment by konsti** [comment]
I have the same problem on **ElasticSearch 0.19.11** with:
- Ruby 1.9.3p327
- Rails 3.2.9
- Tire 0.5.1
- MongoId 3.0.13
- Moped 1.2.9
- Origin 1.0.10

**Comment by digitalplaywright** [comment]
@konsti I have not had time to look debug Tire to figure out what is going wrong here. Have you made any progress on this issue?

@karmi Please let us know if you have any advice on what may be the culprit code in Tire or if you have time to help debug this issue.

**Comment by konsti** [comment]
@digitalplaywright We use this versions in a small side project, so I just don't use bulk import. Didn't find any time to investigate further. I'll keep you posted.

**Comment by sideshow** [comment]
See this pull request karmi/tire#535, which should fix this issue.

**Comment by karmi** [comment]
It seems that _Mongoid_ returns a BSON object, instead of some plain JSON-friendly object for the IDs. See tire#535.

Either we're incorrectly serializing _Mongoid_ objects in _Tire_, or _Mongoid_ needs some special care. Any _Mongoid_ specialist to advice here? //cc @michaelklishin

**Comment by digitalplaywright** [comment]
The Mongoid BSON object is serializable to json. However, to stay consistent with the json from the MongoDB core database itself during export, Mongoid to_json on the id field returns "{\"$oid\": \"object-id-here\"}". It might be this that confuses Tire.

**Comment by karmi** [comment]
I understand that. But when I have a plain `Article` class (with `include Mongoid::Document`), the output of `as_json`, `to_json` and `serializable_hash` looks OK to me:

```
>> a = Article.first
=> #<Article _id: 507d19bc410b5a5d85000001, _type: nil, title: "One", content: "...", published_on: 2012-10-16 08:24:28 UTC>
>> a.as_json
=> {"_id"=>"507d19bc410b5a5d85000001", "content"=>"...", "published_on"=>2012-10-16 08:24:28 UTC, "title"=>"One"}
>> a.to_json
=> "{\"_id\":\"507d19bc410b5a5d85000001\",\"content\":\"...\",\"published_on\":\"2012-10-16T08:24:28Z\",\"title\":\"One\"}"
>> a.serializable_hash
=> {"_id"=>"507d19bc410b5a5d85000001", "content"=>"...", "published_on"=>2012-10-16 08:24:28 UTC, "title"=>"One"}
```

So, what should we do differently in _Tire_?

**Comment by digitalplaywright** [comment]
as_json and to_json directly on the object id seems to be the problem:

```
1.9.3p194 :003 > Moped::BSON::ObjectId.from_time(Time.now).to_json
 => "{\"$oid\": \"50c0d6bd0000000000000000\"}" 
1.9.3p194 :004 > Moped::BSON::ObjectId.from_time(Time.now).as_json
 => "50c0d6d60000000000000000" 
1.9.3p194 :005 > 
```

I am not sure why Mongoid serialize bson ids differently when calling to_json and as_json on the document. It seems like an inconsistency to me.

**Comment by karmi** [comment]
@digitalplaywright Thank you for all the digging! I think the fastest and safest way right now is the `as_json` as outlined in https://github.com/karmi/tire/pull/535#commitcomment-2277626...

**Comment by karmi** [comment]
Pushed [karmi/tire@mongo_id](https://github.com/karmi/tire/tree/mongo_id) with some example code and tests.

The problem is, I'm not being able to replicate the issue with a simple Mongoid Rails application.

My example model looks like this:

``` ruby
class Article
  include Mongoid::Document

  field :title, :type => String
  field :content, :type => String

  include Tire::Model::Search
  include Tire::Model::Callbacks

  def self.paginate(options = {})
     page(options[:page]).per(options[:per_page])
  end

end
```

I can import the data just fine with either `bundle exec rake environment tire:import CLASS='Article' FORCE=true` or `Article.import` in the console, and the IDs in the index correspond to the IDs in the Mongo database...

**Comment by sideshow** [comment]
@karmi see if you can reproduce with `gem "mongoid", "~> 3.0.0"`

**Comment by karmi** [comment]
@sideshow That's the version I have:

```
$ bundle list | grep mongo
mongoid (3.0.14)
```

**Comment by karmi** [comment]
@sideshow @digitalplaywright Would it be possible to upload a simple Rails app which would demonstrate the issue for me? I'm not against pulling the changes from the linked commit, but would rather go without it, unless absolutely neccessary.

**Comment by digitalplaywright** [comment]
@karmi I did some digging and I am only able to reproduce the problem when 'yajl/json_gem' is required  in 'config/application.rb'. This branch demonstrates the bug: 
- https://github.com/digitalplaywright/echo/tree/tire_bug

**Comment by digitalplaywright** [comment]
require 'json/json_gem' overrides the standard Ruby json helpers with the Yajl equivalents:

JSON.parse, JSON.generate, JSON.pretty_generate, JSON.load, JSON.dump and all of the #to_json instance method overrides for Ruby's primitive objects

That is probably why we are seeing this slightly different behavior.

**Comment by karmi** [comment]
@digitalplaywright Many thanks for the perseverance in hunting this bug! I've pushed the `as_json` solution in the attached commit, import now runs fine with your example app (`TestCase.build_new`).

**Comment by digitalplaywright** [comment]
@karmi Thank you for fixing the bug. Please let me know if you are interested in making a release soon with these changes. In the meantime we will use a direct gem dependency with master.

@sideshow @karmi Btw, I would like to create a unit test independently from Tire that demonstrates exactly what is different when you require 'yajl/json_gem'. Please let me know if you have a test case that demonstrates the problem independently of Tire.

**Comment by karmi** [comment]
Released as [0.5.2](https://github.com/karmi/tire/compare/v0.5.1...v0.5.2)

@digitalplaywright I don't have any isolated test with just Yajl...

**Comment by digitalplaywright** [comment]
@karmi Perfect. Thank you!
